Starting from saved state, delay system file creation until ServicesMain#init

**Summary of the defect**
When starting from a saved state, it appears `BinaryObjectStore` recovery is not completed until _after_ `ServicesState#init`. In this case we must delay system file initialization to `ServicesMain#init`---as was done before reonnect.

**Service logs (if applicable)**
```
...
2020-11-18 17:44:58.635 INFO  445  SwirldsPlatform - Signed state loaded from disk has a valid hash.
2020-11-18 17:44:58.689 ERROR 703  Browser - Saved state not loaded:
com.swirlds.platform.internal.SignedStateLoadingException: Exception while reading signed state!
	at com.swirlds.platform.SwirldsPlatform.loadSavedStateFromDisk(SwirldsPlatform.java:473) ~[swirlds-platform-core-0.7.5.jar:?]
	at com.swirlds.platform.Browser.createLocalPlatforms(Browser.java:701) [swirlds-platform-core-0.7.5.jar:?]
	at com.swirlds.platform.Browser.startPlatforms(Browser.java:944) [swirlds-platform-core-0.7.5.jar:?]
	at com.swirlds.platform.Browser.main(Browser.java:345) [swirlds-platform-core-0.7.5.jar:?]
Caused by: java.lang.NullPointerException
	at com.swirlds.blob.BinaryObject.getId(BinaryObject.java:130) ~[swirlds-platform-core-0.7.5.jar:?]
	at com.swirlds.blob.BinaryObjectStore.get(BinaryObjectStore.java:297) ~[swirlds-platform-core-0.7.5.jar:?]
	at com.hedera.services.state.merkle.MerkleOptionalBlob.getData(MerkleOptionalBlob.java:216) ~[?:?]
	at java.util.Optional.map(Optional.java:265) ~[?:?]
	at com.hedera.services.files.store.FcBlobsBytesStore.get(FcBlobsBytesStore.java:111) ~[?:?]
	at com.hedera.services.files.store.FcBlobsBytesStore.get(FcBlobsBytesStore.java:37) ~[?:?]
	at com.hedera.services.files.store.BytesStoreAdapter.get(BytesStoreAdapter.java:79) ~[?:?]
	at com.hedera.services.files.TieredHederaFs.assertUsable(TieredHederaFs.java:326) ~[?:?]
	at com.hedera.services.files.TieredHederaFs.cat(TieredHederaFs.java:136) ~[?:?]
	at com.hedera.services.state.initialization.HfsSystemFilesManager.loadFrom(HfsSystemFilesManager.java:170) ~[?:?]
	at com.hedera.services.state.initialization.HfsSystemFilesManager.loadConfigWithJutilPropsFallback(HfsSystemFilesManager.java:231) ~[?:?]
	at com.hedera.services.state.initialization.HfsSystemFilesManager.loadApplicationProperties(HfsSystemFilesManager.java:123) ~[?:?]
	at com.hedera.services.ServicesState.initializeContext(ServicesState.java:244) ~[?:?]
	at com.hedera.services.ServicesState.init(ServicesState.java:220) ~[?:?]
	at com.swirlds.platform.SwirldsPlatform.loadSavedStateFromDisk(SwirldsPlatform.java:461) ~[swirlds-platform-core-0.7.5.jar:?]
	... 3 more
2020-11-18 17:44:58.700 DEBUG 210  BinaryObjectStore - Recovery Starting
2020-11-18 17:44:58.701 DEBUG 222  BinaryObjectStore - Recovering Objects [ objectCount = 13, uniqueObjectCount = 7 ]
2020-11-18 17:44:58.714 DEBUG 267  BinaryObjectStore - Recovery Finished
...
```

--- a/hedera-node/src/main/java/com/hedera/services/ServicesMain.java
+++ b/hedera-node/src/main/java/com/hedera/services/ServicesMain.java
@@ -185,8 +185,10 @@ public class ServicesMain implements SwirldMain {
 		try {
 			ctx.systemFilesManager().createAddressBookIfMissing();
 			ctx.systemFilesManager().createNodeDetailsIfMissing();
-			ctx.systemFilesManager().loadExchangeRates();
 			ctx.systemFilesManager().createUpdateZipFileIfMissing();
+			if (!ctx.systemFilesManager().areFilesLoaded()) {
+				ctx.systemFilesManager().loadAllSystemFiles();
+			}
 		} catch (Exception e) {
 			throw new IllegalStateException("Could not create system files!", e);
 		}

--- a/hedera-node/src/main/java/com/hedera/services/ServicesState.java
+++ b/hedera-node/src/main/java/com/hedera/services/ServicesState.java
@@ -39,6 +39,7 @@ import com.hedera.services.state.submerkle.ExchangeRates;
 import com.hedera.services.state.submerkle.SequenceNumber;
 import com.hedera.services.utils.PlatformTxnAccessor;
 import com.hederahashgraph.api.proto.java.AccountID;
+import com.swirlds.blob.BinaryObjectStore;
 import com.swirlds.common.Address;
 import com.swirlds.common.AddressBook;
 import com.swirlds.common.NodeId;
@@ -80,6 +81,7 @@ public class ServicesState extends AbstractMerkleInternal implements SwirldState
 
 	static Consumer<MerkleNode> merkleDigest = CryptoFactory.getInstance()::digestTreeSync;
 	static Supplier<AddressBook> legacyTmpBookSupplier = AddressBook::new;
+	static Supplier<BinaryObjectStore> blobStoreSupplier = BinaryObjectStore::getInstance;
 
 	NodeId nodeId = null;
 	boolean skipDiskFsHashCheck = false;
@@ -236,21 +238,9 @@ public class ServicesState extends AbstractMerkleInternal implements SwirldState
 		* history. This history has two main uses: Purging expired records, and
 		* classifying duplicate transactions. */
 		ctx.recordsHistorian().reviewExistingRecords();
-
-		/* Ensure files 0.0.121 and 0.0.122 exist in state, creating them from
-		* the application.properties and api-permission.properties assets if they
-		* are missing. (The {@code SystemFilesManager} will signal interested
-		* components of the loaded files via a callback.) */
-		ctx.systemFilesManager().loadApplicationProperties();
-		ctx.systemFilesManager().loadApiPermissions();
-
-		/* Ensure files 0.0.111 and 0.0.112 exist in state, creating them
-		* from the FeeSchedules.json and bootstrap.properties resources/files
-		* if they are missing. (The {@code SystemFilesManager} will signal interested
-		* components of the loaded files via a callback.) */
-		ctx.systemFilesManager().loadFeeSchedules();
-
-		ctx.systemFilesManager().loadExchangeRates();
+		if (!blobStoreSupplier.get().isInitializing()) {
+			ctx.systemFilesManager().loadAllSystemFiles();
+		}
 	}
 
 	@Override

--- a/hedera-node/src/main/java/com/hedera/services/state/initialization/HfsSystemFilesManager.java
+++ b/hedera-node/src/main/java/com/hedera/services/state/initialization/HfsSystemFilesManager.java
@@ -65,7 +65,7 @@ public class HfsSystemFilesManager implements SystemFilesManager {
 	private static final String FEE_SCHEDULES_TAG = "fee schedules";
 
 	private JKey systemKey;
-
+	private boolean filesLoaded = false;
 	private final AddressBook currentBook;
 	private final FileNumbers fileNumbers;
 	private final PropertySource properties;
@@ -145,6 +145,16 @@ public class HfsSystemFilesManager implements SystemFilesManager {
 				schedulesCb,
 				CurrentAndNextFeeSchedule::parseFrom,
 				() -> defaultSchedules().toByteArray());
+	}
+
+	@Override
+	public void setFilesLoaded() {
+		filesLoaded = true;
+	}
+
+	@Override
+	public boolean areFilesLoaded() {
+		return filesLoaded;
 	}
 
 	@Override

--- a/hedera-node/src/main/java/com/hedera/services/state/initialization/SystemFilesManager.java
+++ b/hedera-node/src/main/java/com/hedera/services/state/initialization/SystemFilesManager.java
@@ -25,9 +25,28 @@ public interface SystemFilesManager {
 	void createNodeDetailsIfMissing();
 	void createUpdateZipFileIfMissing();
 
+	/* Ensure files 0.0.111 and 0.0.112 exist in state, creating them
+	 * from the FeeSchedules.json and bootstrap.properties resources/files
+	 * if they are missing. (The {@code HfsSystemFilesManager} will signal interested
+	 * components of the loaded files via a callback.) */
 	void loadExchangeRates();
 	void loadFeeSchedules();
 
+	/* Ensure files 0.0.121 and 0.0.122 exist in state, creating them from
+	 * the application.properties and api-permission.properties assets if they
+	 * are missing. (The {@code HfsSystemFilesManager} will signal interested
+	 * components of the loaded files via a callback.) */
 	void loadApiPermissions();
 	void loadApplicationProperties();
+
+	default void loadAllSystemFiles() {
+		loadApplicationProperties();
+		loadApiPermissions();
+		loadFeeSchedules();
+		loadExchangeRates();
+
+		setFilesLoaded();
+	}
+	void setFilesLoaded();
+	boolean areFilesLoaded();
 }

--- a/hedera-node/src/test/java/com/hedera/services/ServicesMainTest.java
+++ b/hedera-node/src/test/java/com/hedera/services/ServicesMainTest.java
@@ -351,16 +351,35 @@ public class ServicesMainTest {
 		// then:
 		verify(grpc).start(intThat(i -> i == 50212), intThat(i -> i == 50213), any());
 	}
-	
+
 	@Test
-	public void managesSystemFiles() {
+	public void loadsSystemFilesIfNotAlreadyDone() {
+		given(systemFilesManager.areFilesLoaded()).willReturn(true);
+
 		// when:
 		subject.init(null, new NodeId(false, NODE_ID));
 
 		// then:
 		verify(systemFilesManager).createAddressBookIfMissing();
 		verify(systemFilesManager).createNodeDetailsIfMissing();
-		verify(systemFilesManager).loadExchangeRates();
+		verify(systemFilesManager).createUpdateZipFileIfMissing();
+		// and:
+		verify(systemFilesManager, never()).loadAllSystemFiles();
+	}
+
+	@Test
+	public void managesSystemFiles() {
+		given(systemFilesManager.areFilesLoaded()).willReturn(false);
+
+		// when:
+		subject.init(null, new NodeId(false, NODE_ID));
+
+		// then:
+		verify(systemFilesManager).createAddressBookIfMissing();
+		verify(systemFilesManager).createNodeDetailsIfMissing();
+		verify(systemFilesManager).createUpdateZipFileIfMissing();
+		// and:
+		verify(systemFilesManager).loadAllSystemFiles();
 	}
 
 	@Test

--- a/hedera-node/src/test/java/com/hedera/services/ServicesStateTest.java
+++ b/hedera-node/src/test/java/com/hedera/services/ServicesStateTest.java
@@ -53,6 +53,7 @@ import com.hederahashgraph.api.proto.java.SignatureMap;
 import com.hederahashgraph.api.proto.java.SignaturePair;
 import com.hederahashgraph.api.proto.java.SignedTransaction;
 import com.hederahashgraph.api.proto.java.TransactionID;
+import com.swirlds.blob.BinaryObjectStore;
 import com.swirlds.common.Address;
 import com.swirlds.common.AddressBook;
 import com.swirlds.common.NodeId;
@@ -79,6 +80,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.function.Consumer;
+import java.util.function.Supplier;
 
 import static com.hedera.services.context.SingletonContextsManager.CONTEXTS;
 import static java.util.Collections.EMPTY_LIST;
@@ -99,6 +101,8 @@ import static org.mockito.BDDMockito.verify;
 @RunWith(JUnitPlatform.class)
 class ServicesStateTest {
 	Consumer<MerkleNode> mockDigest;
+	Supplier<BinaryObjectStore> mockBlobStoreSupplier;
+	BinaryObjectStore blobStore;
 	Instant now = Instant.now();
 	Transaction platformTxn;
 	Address address;
@@ -139,6 +143,11 @@ class ServicesStateTest {
 		CONTEXTS.clear();
 		mockDigest = (Consumer<MerkleNode>)mock(Consumer.class);
 		ServicesState.merkleDigest = mockDigest;
+		blobStore = mock(BinaryObjectStore.class);
+		mockBlobStoreSupplier = (Supplier<BinaryObjectStore>) mock(Supplier.class);
+		given(mockBlobStoreSupplier.get()).willReturn(blobStore);
+		ServicesState.blobStoreSupplier = mockBlobStoreSupplier;
+		given(blobStore.isInitializing()).willReturn(false);
 
 		out = mock(SerializableDataOutputStream.class);
 		in = mock(SerializableDataInputStream.class);
@@ -203,6 +212,7 @@ class ServicesStateTest {
 
 		subject = new ServicesState();
 	}
+
 
 	@Test
 	void ensuresNonNullTokenFcmsAfterReadingFromLegacySavedState() {
@@ -287,10 +297,27 @@ class ServicesStateTest {
 		inOrder.verify(ctx).update(subject);
 		inOrder.verify(ctx).rebuildBackingStoresIfPresent();
 		inOrder.verify(historian).reviewExistingRecords();
-		inOrder.verify(systemFilesManager).loadApplicationProperties();
-		inOrder.verify(systemFilesManager).loadApiPermissions();
-		inOrder.verify(systemFilesManager).loadFeeSchedules();
-		inOrder.verify(systemFilesManager).loadExchangeRates();
+		inOrder.verify(systemFilesManager).loadAllSystemFiles();
+	}
+
+	@Test
+	public void doesntInitializeFilesIfStoreStillInitializing() {
+		InOrder inOrder = inOrder(ctx, txnHistories, historian, systemFilesManager);
+
+		given(blobStore.isInitializing()).willReturn(true);
+		given(ctx.nodeAccount()).willReturn(AccountID.getDefaultInstance());
+		// and:
+		CONTEXTS.store(ctx);
+
+		// when:
+		subject.init(platform, book);
+
+		// then:
+		inOrder.verify(ctx).nodeAccount();
+		inOrder.verify(ctx).update(subject);
+		inOrder.verify(ctx).rebuildBackingStoresIfPresent();
+		inOrder.verify(historian).reviewExistingRecords();
+		inOrder.verify(systemFilesManager, never()).loadAllSystemFiles();
 	}
 
 	@Test
@@ -692,5 +719,6 @@ class ServicesStateTest {
 	public void cleanup() {
 		CONTEXTS.clear();
 		ServicesState.merkleDigest = CryptoFactory.getInstance()::digestTreeSync;
+		ServicesState.blobStoreSupplier = BinaryObjectStore::getInstance;
 	}
 }

--- a/hedera-node/src/test/java/com/hedera/services/state/initialization/HfsSystemFilesManagerTest.java
+++ b/hedera-node/src/test/java/com/hedera/services/state/initialization/HfsSystemFilesManagerTest.java
@@ -204,6 +204,36 @@ class HfsSystemFilesManagerTest {
 	}
 
 	@Test
+	public void loadsEverything() {
+		// setup:
+		SystemFilesManager sub = mock(SystemFilesManager.class);
+
+		willCallRealMethod().given(sub).loadAllSystemFiles();
+
+		// when:
+		sub.loadAllSystemFiles();
+
+		// then:
+		verify(sub).loadApplicationProperties();
+		verify(sub).loadApiPermissions();
+		verify(sub).loadFeeSchedules();
+		verify(sub).loadExchangeRates();
+		verify(sub).setFilesLoaded();
+	}
+
+	@Test
+	public void tracksFileLoading() {
+		// expect:
+		assertFalse(subject.areFilesLoaded());
+
+		// when:
+		subject.setFilesLoaded();
+
+		// then:
+		assertTrue(subject.areFilesLoaded());
+	}
+
+	@Test
 	public void doesntCreateAddressBookIfPresent() {
 		given(hfs.exists(bookId)).willReturn(true);
 

--- a/test-clients/devops-utils/validation-scenarios/config.yml
+++ b/test-clients/devops-utils/validation-scenarios/config.yml
@@ -92,12 +92,12 @@ networks:
     - {account: 3, ipv4Addr: '127.0.0.1:50211'}
     scenarioPayer: 1001
     scenarios:
-      feeSnapshots:
-        appendToSnapshotCsv: false
-        ignoreCostAnswer: true
-        opsConfig: {bytecode: 1002, memoLength: 32}
-        scheduleDesc: Bootstrap
-        tinyBarsToOffer: 10000000000
+      consensus: {persistent: 1007}
+      contract:
+        persistent: {bytecode: 1005, luckyNo: 42, num: 1006, source: Multipurpose.sol}
+      crypto: {receiver: 1003, sender: 1002}
+      file:
+        persistent: {contents: MrBleaney.txt, num: 1004}
       sysFilesDown:
         evalMode: snapshot
         numsToFetch: [111]

